src/tallies/filter.cpp

```diff
@@ -0,0 +1,249 @@
+#include "openmc/tallies/filter.h"
+
+#include <algorithm> // for max
+#include <cstring> // for strcpy
+#include <string>
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"  // for MAX_LINE_LEN;
+#include "openmc/error.h"
+#include "openmc/xml_interface.h"
+#include "openmc/tallies/filter_azimuthal.h"
+#include "openmc/tallies/filter_cell.h"
+#include "openmc/tallies/filter_cellborn.h"
+#include "openmc/tallies/filter_cellfrom.h"
+#include "openmc/tallies/filter_delayedgroup.h"
+#include "openmc/tallies/filter_distribcell.h"
+#include "openmc/tallies/filter_energyfunc.h"
+#include "openmc/tallies/filter_energy.h"
+#include "openmc/tallies/filter_legendre.h"
+#include "openmc/tallies/filter_material.h"
+#include "openmc/tallies/filter_mesh.h"
+#include "openmc/tallies/filter_meshsurface.h"
+#include "openmc/tallies/filter_mu.h"
+#include "openmc/tallies/filter_particle.h"
+#include "openmc/tallies/filter_polar.h"
+#include "openmc/tallies/filter_sph_harm.h"
+#include "openmc/tallies/filter_sptl_legendre.h"
+#include "openmc/tallies/filter_surface.h"
+#include "openmc/tallies/filter_universe.h"
+#include "openmc/tallies/filter_zernike.h"
+
+// explicit template instantiation definition
+template class std::vector<openmc::FilterMatch>;
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace simulation {
+  std::vector<FilterMatch> filter_matches;
+}
+
+namespace model {
+  std::vector<std::unique_ptr<Filter>> tally_filters;
+  std::unordered_map<int, int> filter_map;
+}
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+extern "C" size_t tally_filters_size()
+{
+  return model::tally_filters.size();
+}
+
+//==============================================================================
+// Filter implementation
+//==============================================================================
+
+Filter::Filter() : index_{model::tally_filters.size()}
+{ }
+
+Filter::~Filter()
+{
+  model::filter_map.erase(id_);
+}
+
+Filter* Filter::create(pugi::xml_node node)
+{
+  // Copy filter id
+  if (!check_for_node(node, "id")) {
+    fatal_error("Must specify id for filter in tally XML file.");
+  }
+  int filter_id = std::stoi(get_node_value(node, "id"));
+
+  // Convert filter type to lower case
+  std::string s;
+  if (check_for_node(node, "type")) {
+    s = get_node_value(node, "type", true);
+  }
+
+  // Allocate according to the filter type
+  auto f = Filter::create(s, filter_id);
+
+  // Read filter data from XML
+  f->from_xml(node);
+  return f;
+}
+
+Filter* Filter::create(const std::string& type, int32_t id)
+{
```

Comment:
It may be cleaner to use a template or a macro instead of repeating the same code. For example
```
template<T> bool tryFilter(const std::string &name, const std::string &type) {
    if(name == type) {
        model::tally_filters.push_back(std::make_unique<T>());
        return(true);
    }
    return(false);
}
Filter* Filter::create(const std::string& type, int32_t id) {
    bool success = tryFilter<AzimuthalFilter>("azimuthal")
                             || tryFilter<CellFilter>("cell")
                             || ...;
    ...
}
```